src/tmpfiles/tmpfiles.c

```diff
@@ -208,20 +216,148 @@ static const Specifier specifier_table[] = {
         {}
 };
 
+static int specifier_kernel_isol(char specifier, const void *data, const void *userdata, char **ret) {
+        if (!empty_or_root(arg_root))
+                return -ENXIO;
+
+        return specifier_kernel_release(specifier, data, userdata, ret);
+}
+
+static int specifier_arch_isol(char specifier, const void *data, const void *userdata, char **ret) {
+        if (!empty_or_root(arg_root))
+                return -ENXIO;
```

Comment:
I think these two should always work the same way. i.e. i think it's something to document in tmpfiles.d man page instead of prohibiting it altogether... i.e. we should just say `%a always resolves to the host's architecture, regardless if '--root=' is used...`  or so

Reply:
Depends on what `--root=` is trying to archive. I would like to build a root filesystem for another architecture and pre-create all files & folders (outside of volatile paths like /tmp and /run). If its the same arch I could execute `systemd-tmpfiles` in a chroot.
Needless to say I would rather have the specifiers missing than wrong. I guess you are mostly thinking about containers for the same machine?
What is `--root=` used for currently?

Reply:
maybe use an environment variable or a separate command line argument to override it ?

Reply:
@nolange but the point I am making is that your patch doesn't make it work either, it just causes the stuff to fail.

I think @boucman has a point: let's maybe make this overridable via an env var, and independent of arg_root or no arg_root. i.e. if `$SYSTEMD_ARCHITECTURE=` is set, then %a would resolve to that rather than the actual architecture. Similar for the other fields.

Reply:
> What is --root= used for currently?

Mostly testing I guess: i.e. run some stuff without affecting the system things are run on.

I mean, I sympathize with the goal of this patch, and I am convinced for the specifiers that are sourced from files of the OS tree we should also operate relative to the same root specified by `--root=`, but this doesn't really work so well for %a and the like, that expose runtime properties more than actual properties of the directory tree. For those I think a better approach is to just inherit from the host, and maybe make this overridable via env var.

Reply:
> @nolange but the point I am making is that your patch doesn't make it work either, it just causes the stuff to fail.

**Decisively** fail, just like there should be no boot-uuid. Its well-defined that the entries containing those specifiers will be skipped, those entries will be created when the relevant specifiers are available ("real" boot).

In the issue report  #16183 I noted the option to add overrides, and this is the path I would like to take.
The safe default however would be leave out specifiers that are not unambiguously available.

Reply:
I really wouldn't blur the line between runtime parameters (i.e. %a, %b, %H, %v, …) and properties of the disk image itself (i.e. %B, %w, %m, …). They are quite different in nature, the former are derived from all kinds of things, but the latter only from the image itself. I wouldn't try to be too smart here, beacuse the runtime parameters are only very very indirectly connected to the disk image, hence I would say that anything is good enough for them, and if people want explicit control over runtime parameters when doing "offline" operation, then env vars are the way to go for those. I don#t think there's value in prohibiting using the current system's runtime params for --root= operation however, because for our test cases that would break stuff...

I mean, you never know what people use --root= for. it could be that they do "staged" stuff, i.e. apply things into a subdir to the copy/merge it into the host fs. In such a case it is totally OK if the runtime specifiers resolve like they normally do.

Hence again: the arg_root stuff should affect %B, %w, %m, absolutely, happy to merge a patch for that. But for %a, %b %H I think a concept of making them overridable via env vars is the better approach is it permits folks who want to set this up for full offline operation to do what they want, but it doesn't limit anything for anyone else.

Reply:
> In such a case it is totally OK if the runtime specifiers resolve like they normally do.

Or you might accidentally leak some information from your host, if you stage a build for a different machine. I'd stay on the side of caution, but I definitely would need **one** switch to block all current and future specifiers.

Ill plan to do the following the next days:

- change the default to pick runtime params from the host
- add a switch `--disable-runtime-specifiers` to change the default to how I (first) implemented it: return -ENXIO, skip affected lines.
- add argument `--set-specifier={boot-id,hostname,architecture,kernel-release}=VALUE` for replacing the values.
  potentially need some special `%host%` and `%null%` VALUES to only switch few values.

Reply:
i think controlling this via env vars makes sense, but I'd prefer not to have high-level switches. "override" style stuff we generally do via env vars, most options in docs/ENVIRONMENT.md are override style stuff. I think we should do this the same way here. There's no shame in using env vars like that.

also, if possible don't do "negative" options, i.e. "--disable-foo"... turning something off by turning it on it not the prettiest thing I believe. This is only a request though, and the threshold for a solution is low. for example "mask" or "prohibit" or "refuse" might already be enough to use instead of "disable" to make me shut up about this. Better though is making it positive. 

Maybe do:

```
$SYSTEMD_SPECIFIER_BOOT_ID=…
$SYSTEMD_SPECIFIER_HOSTNAME=…
$SYSTEMD_SPECIFIER_PERMIT_RUNTIME=<bool>
…
```

Reply:
I think I leave this open, as this is the main thread of discussion, I am not done yet and I will go over the formatting later. Got little spare time now, just tell me if I am on the right track so far.

The tool now works identical as before, until you set `SYSTEMD_SPECIFIER_PROHIBIT_RUNTIME`, which is intended to
-   do the same thing as the `systemd-tmpfiles-setup.service` on a `--root=prefix`. But take into account that this might be a different machine than the host
-   specifiers that cant be unambiguous resolved should cause the entry to be skipped (need to thing of an better warning msg there).
-   some `SYSTEMD_SPECIFIER_` env variables should allow setting a few specifiers directly (TODO)